### DISCUSSION TITLE
Inconsistent add_api_route types

### DISCUSSION THREAD

**Discussion by sidekick-eimantas** [fastapi/fastapi#10235] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
### Direct mounting of routes
import pathlib

from fastapi import FastAPI, APIRouter
import uvicorn
import pydantic

class HelloRequest(pydantic.BaseModel):
    id: str


class HelloResponse(pydantic.BaseModel):
    hello: str


class Hello:
    def __init__(self, name: str) -> None:
        self.name = name

    async def handle(self, request: HelloRequest) -> HelloResponse:
        return HelloResponse(hello=self.name)

app = FastAPI()
hello_handler = Hello(name="test")
app.add_api_route("/hello", hello_handler.handle, methods=["POST"])






### Mounting of paths via Router
import pathlib

from fastapi import FastAPI, APIRouter
import uvicorn
import pydantic

class HelloRequest(pydantic.BaseModel):
    id: str


class HelloResponse(pydantic.BaseModel):
    hello: str


class Hello:
    def __init__(self, name: str) -> None:
        self.name = name

    async def handle(self, request: HelloRequest) -> HelloResponse:
        return HelloResponse(hello=self.name)


class RootRouter:
    def __init__(self, name: str):
        self.router = APIRouter()
        hello_handler = Hello(name=name)
        self.router.add_api_route("/hello", hello_handler.handle, methods=["POST"])


app = FastAPI()
root_router = RootRouter(name="test")
app.include_router(root_router.router)
```

### Description

First example produces a mypy error:
```
error: Argument 2 to "add_api_route" of "FastAPI" has incompatible type "Callable[[HelloRequest], Coroutine[Any, Any, HelloResponse]]"; expected "Callable[..., Coroutine[Any, Any, Response]]"  [arg-type]
```

Second example does not.

The types of `endpoint` parameter in `FastAPI.add_api_route` and `APIRouter.add_api_route` are inconsistent.

### Operating System

macOS

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.7

**Comment by Kludex** [comment, marked as answer]
As much as simple, and trivial this is to fix, it's a valid thing to fix. 😅 

Thanks.

**Reply by sidekick-eimantas** [reply]
Happy to raise a PR for this - wasn't sure whether this was intended or changing it had wider implications

**Reply by Kludex** [reply]
I don't know which of those type hints is wanted. 🤷‍♂️

I've created https://github.com/tiangolo/fastapi/issues/10236.

**Reply by tiangolo** [reply]
Thanks! This was handled here: https://github.com/fastapi/fastapi/pull/10240 :rocket:
